[SKU Scanner] Add support for scanning variations

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of https://github.com/woocommerce/woocommerce-ios/issues/9658
Closes: #9657 

## Description
This PR adds support for scanning product variations from the 2 scanning entry points: 
  * Order creation: The scanning icon in the Order List navigation bar
  * Order editing: The `+ Add Product via SKU` button within an Editable Order.

Caveats:
- Scanning the same product more than once (Order List view, and then Order Editable view) will add it more than once as different rows.
- Sometimes I've seen an UI glitch that shows the parent product rather than the variation being selected on Order Editing, this resolves automatically when we tap on "create" order (the correct one was scanned and passed to the order). I'll research this further on a different PR.

## Changes
- Adds a `syncOrderWithProduct(productID)` method, that validates if the passed ID belongs to a product, or a product variation by comparing what we've scanned with what's in local storage. If there's a match, we send the specific type to the order synchronizer. Otherwise, we'll display an error notice.
- ~~Removed `test_addScannedProductToOrder_when_existing_sku_is_found_then_succeeds_to_add_product_to_order` test temporarily. More details here: p1684490283255659-slack-C025A8VV728, logged [here](https://github.com/woocommerce/woocommerce-ios/issues/9771)~~

## Testing instructions
Have a product variation with a decoded SKU in your store by creating a barcode in [this website](https://www.barcode-generator.org/) > open the app > Products > Select a product with variations > Tap on Variations > Tap on a specific variation > Inventory > Tap on the barcode icon > scan the barcode created previously > tap done > tap save.

**Case 1: Order Creation**
- Run the app > go to Orders > Tap the scanner icon > and scan the variation barcode
- See that the scanner is dismissed, and we start the order creation flow with the product variation already added to the Order.

**Case 2: Order Editing**
- In the same or a different Order being edited, tap the `+ Add Product via SKU`  > scan the variation barcode
- See that the scanner is dismissed, and the product variation has been added to the order

~~Case 3: Unhappy path~~

~~This case handles when a SKU product or variation match can be found in the API, but we couldn't match it to any product or variation saved locally. The easiest way to test this is by commenting out line 949 and calling the notice instead:~~ 

## Screenshots
| Scan on Order creation | Scan on Order editing | Error notice |
|--------|--------|--------|
| <img width=450 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/1c6f2ce7-49b4-4608-b458-76e313856b55"> | <img width=450 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/0c91f5c6-de05-4aa5-8f47-ac0bb0465a8b"> |  <img width=450 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/acce0492-464d-47d2-b554-b82e8dab48eb"> |

iamgabrielma committed:
Validate scanned ID as product or productVariation

iamgabrielma committed:
Add unit tests for variation scanning happy path

iamgabrielma committed:
Add error notice. Unit Test fail path

iamgabrielma committed:
Extract product or variation validation to a common method

As the same logic is used from two different flows (order creation, order editing), we extract the validation method to be used when scanning products from the different views

iamgabrielma committed:
Remove unit test temporarily

Since we check the contents of allProducts array when filtering if we should send a product or variation to the ordersynchronizer, and this array rely on resultscontroller to be filled, which happens when configuring the different product row models the, test fails as allProducts is empty when we attempt to access it.

iamgabrielma committed:
Rename method

iamgabrielma commented:
One review is enough @joshheald @toupper , just added you both because if almost weekend and Josh will be AFK on Monday. There's no rush in any case, as is targeting 13.8 🙇

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -1271,6 +1265,7 @@ extension EditableOrderViewModel {
     enum ScannerError: Error {
         case nilSKU
         case productNotFound
+        case productNotFoundInLocalStorage
```
I am wondering, why is a Scanner Error when the product is not found in local storage? More than an error, I imagine that that's a common scenario and I wouldn't call that an error :)

iamgabrielma replied:
Agree, removed on https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/45b3cb045c98a0837b202f8d6dc79349a9825fd1

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -1339,6 +1348,12 @@ extension EditableOrderViewModel {
 
+        /// Returns an error notice when the given product ID cannot be found in local storage
+        ///
+        static func createOrderFromScannedProductErrorNotice(for ID: Int64) -> Notice {
```
This is related to the comment above, why do we show an error if the product cannot be found in local storage? I don't think neither that's per se an error, nor we should communicate it to the user.

iamgabrielma replied:
It seems not needed indeed, specially since the latest refactor we expect to have a product or variation when this point is reached. The only error that could happen at this point is within the synchronizer, and we already have an implementation in place that will deal with this. I removed the method and the error case in the latests commits.

iamgabrielma replied:
Agree, removed on https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/c38d3a4d644d3cb82da3c0f398b74a205cda1389

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -1323,6 +1318,20 @@ extension EditableOrderViewModel {
+    private func updateOrderWithProduct(_ productID: Int64) {
+        if let product = allProducts.first(where: { $0.productID == productID }) {
+            orderSynchronizer.setProduct.send(.init(product: .product(product), quantity: 1))
+        } else if let productVariation = allProductVariations.first(where: { $0.productVariationID == productID }) {
```
I find it a bit confusing that we are naming the function `updateOrderWithProduct` and the parameter `productID` but we also handle variations and variation ids, could we perhaps name it in a more generic way to handle product variations as well?

iamgabrielma replied:
As we changed the signature to `updateOrderWithProduct(_ product: Product)` in https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/c38d3a4d644d3cb82da3c0f398b74a205cda1389 , I left the name as it is, but happy to change it to something else if you think should be clearer 👍

toupper commented:
@iamgabrielma Thanks for the PR!

I've just tested it, and the first case (Case 1: Order Creation) doesn't work as expected when added for the first time. I scan the code successfully, but when the order is created it's empty. After I create the order and open it again the product appears:


https://github.com/woocommerce/woocommerce-ios/assets/1864060/88e0dea0-4d06-414b-937a-44e996e971de

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/620790e3-7f8c-4b20-813b-0578ad41d5e6" alt="After the order is created the item is there" width="375">

iamgabrielma committed:
Pass product entity to EditableOrder rather than only the ID

iamgabrielma committed:
Extract OrderRoot navigation logic to common method

iamgabrielma committed:
Adapt Unit Tests to updated method signature

iamgabrielma committed:
Remove case for productNotFoundInLocalStorage scanning error

iamgabrielma commented:
Thanks for testing! I believe the issue you've seen could be related to this: p1684490283255659-slack-C025A8VV728 , when we send the item to the order synchronizer, the product rows may not be rendered on time as depends on whatever we have on local storage, rather than on the product we've just fetched. While we upsert results upon SKU match, we still may have the case where this is not rendered because is not sharing the same source.

As discussed in our call I've changed the `mapFromSKUtoProduct` signature to return the Product entity we get from the network, rather than the productID and then attempt to find it in the `allProducts` array. Let me know if that resolves the issue!

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -940,19 +941,13 @@ private extension EditableOrderViewModel {
     }
 
-    /// If given, sends a product to the Order synchronizer during the initial Order creation setup
+    /// If given an initial item during the Order creation setup, sync this product with the Order
```
```suggestion
    /// If given an initial Product or ProductVariation on initialization, updates the Order with the entity
```

iamgabrielma replied:
https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/12f22b15c7a21cf86a68867026f65191b23d3fca

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift:
```diff
@@ -155,18 +155,27 @@ final class OrdersRootViewController: UIViewController {
-    /// and validate it when needed.
+    @objc func presentOrderCreationFlow() {
+        let viewModel = EditableOrderViewModel(siteID: siteID)
+        let viewController = OrderFormHostingController(viewModel: viewModel)
```
Perhaps I can create the viewController inside the `setupNavigation` method as well, rather than pass it as a property. I'll test this further separately as we're just saving one duplicated line here.

iamgabrielma replied:
Changed here: https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/cab6b3c45292897845f1476a90bf9ef484149c1a

toupper commented:
@iamgabrielma I tried again and am still able to reproduce the issue described above. Now we don't even show the rows loading when the order is opened:


https://github.com/woocommerce/woocommerce-ios/assets/1864060/8f03a512-6732-4e73-8af3-8f2c832ffa2e

As you can see, when the order is created the rows are shown properly.

iamgabrielma committed:
Move viewcontroller setup to a different method

iamgabrielma committed:
Update documentation

iamgabrielma committed:
Merge branch 'trunk' into issue/9657-support-for-scanning-variations

iamgabrielma commented:
@toupper can you confirm if this happens only for variations? Or for products as well? 

Every time I was able to replicate this behaviour, the scenario seems to be "_when scanning a variation for first time, if we have not fetched yet its parent product_". 

Once I tap in the Products tab, I see what we scanned in that view as we already have upserted these to local storage, then immediately after, we load the rest of products. After this, I seem to be able to scan the variation right away. 

Here are two barcodes to try with [the same testing site](https://indiemelon.mystagingwebsite.com/):

| Product | Product Variation |
|--------|--------|
| <img width=450 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/7d760cc1-cf1d-4d73-a696-397301bb428e"> | <img width=450 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/a0d3cceb-b00b-4020-9028-74151747d3d5"> |

iamgabrielma committed:
Handle product row creation for variations when no parent product in storage

iamgabrielma commented:
The issue you've seen should be resolved with https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/eda0a7e41429bf68860bf8b16860418b74588902. This PR is ready for re-review!

Previously, when generating the [ProductRowViewModel](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/EditableOrderViewModel.swift#L437) for variations, we've always assumed that there would be a parent product in local storage, but since this may not be the case anymore now that we're adding variations directly from Order List without fetching first, we were returning early and not creating this product row:
```
let product = allProducts.first(where: { $0.productID == item.productID }) else {
  return nil
}
```

An easier way to test this behaviour is to reset local storage every time `OrdersRootViewController` is loaded by calling the `resetStoredProducts()` action in, for example, `viewWillAppear`:

```
override func viewWillAppear(_ animated: Bool) {
  super.viewWillAppear(animated)
  resetLocalStorage()
}

func resetLocalStorage() {
  let action = ProductAction.resetStoredProducts(onCompletion: {
    print("All products deleted")
  })
  ServiceLocator.stores.dispatch(action)
}
```
By removing the early-return in favor of checking if a scanned product could be a product or a variation, we now generate the correct rows no matter what entity is scanned.

iamgabrielma committed:
Check for variations before checking products

toupper commented:
@iamgabrielma After the changes it still happens:


https://github.com/woocommerce/woocommerce-ios/assets/1864060/0c1381d5-e62c-4a26-b538-820760e89833

toupper committed:
Look for variations in the products array as well as the remote request might return variations as products

iamgabrielma committed:
Extract `retrieveVariation` to a separate method

iamgabrielma commented:
Thanks for the call @toupper , I've extracted your commit into a separate method and moved it to a private extension. After some more digging, we found an issue with the `searchProducts` action, when upserting results to local storage, which results in variations being saved as product models. I've opened an issue [here](https://github.com/woocommerce/woocommerce-ios/issues/9808) in order to debug this further, and plan a long term solution.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -318,9 +318,10 @@ final class EditableOrderViewModel: ObservableObject {
+    /// It can be either a Product or ProductVariation entity
     ///
-    private let withInitialProductID: Int64?
+    private let withInitialProduct: Yosemite.Product?
```
Now that we see that passing the whole product is not necessary because we are retrieving it from the local storage anyways, I would leave it by passing the ID with Int64 type. As we can pass either product or variation ids, I would rename it to something more generic such as `initialItemID` (without the prefix `with`, which is not required for the property.

iamgabrielma replied:
Sounds good. I switched back to use product IDs on https://github.com/woocommerce/woocommerce-ios/pull/9768/commits/f5cb4a9d3a3fa85c8eb8e5e38e9bba19a6ee7c0e

toupper reviewed (approved):
Thanks for the changes @iamgabrielma! Now the variations and products are properly added. One thing I observed is that when scanning variations the image is not shown in the row:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/48a26009-69e4-484a-99ef-3b4a06ed8fc6" width="375">


But I don't know if it's caused by changes in this PR. Apart from that I just left a comment about the code. Good work! :shipit: 🚢

iamgabrielma committed:
Revert to use product IDs rather than models for scan setup

iamgabrielma commented:
> One thing I observed is that when scanning variations the image is not shown in the row:

Yes, that's an existing issue for product subtypes, some examples: 
- https://github.com/woocommerce/woocommerce-ios/issues/3733
- https://github.com/woocommerce/woocommerce-ios/issues/6560

My hunch is that could be related of what we've seen regarding how we upsert variations or other types to local storage, as attributes for the variation could go missing if we're storing it as a product, but I haven't looked into it.
